include/gp_stack.h

```diff
@@ -0,0 +1,69 @@
+//SPDX-License-Identifier: LGPL-2.0-or-later
+/* Copyright (C) 2020 Richard Palethorpe <richiejp@f-m.fm> */
+
+#include <stddef.h>
+#include <core/gp_common.h>
+
+#ifndef GP_STACK_H__
+#define GP_STACK_H__
+
+/* Get a pointer to the outer gp_stack struct from a pointer to the (start of)
+ * its data.
+ */
+#define GP_STACK(ptr) (GP_CONTAINER_OF(ptr, gp_stack, payload))
+
+/* Get the number of elements in a stack from a pointer to its data. */
+#define GP_STACK_LEN(ptr) (GP_STACK(ptr)->length)
+
+/* Insert a gap into a stack, which is refered to by a pointer to its data ptr.
+ * See gp_stack_gap().
+ */
+#define GP_STACK_GAP(ptr, i, len) \
+	((typeof(ptr))gp_stack_gap(GP_STACK(ptr), i, len))
+
+/* Free the stack containing ptr. */
+#define GP_STACK_FREE(ptr) (free(GP_STACK(ptr)))
+
```

Comment:
I guess that if the code is structured as it it, i.e. the payload is part of the same structure there is no point in having these macros and we can make the functions take void pointer instead and deal with the offset inside of the library code.

And the rest of the trivial functions can be static inline and put into the header.

Also please make the free() no-op on NULL pointers just like any other free() function.

e.g.
```
static inline void gp_stack_free(void *self)
{
         if (!self)
                    return;
         free(GP_STACK(self));
}
```